Production_scripts/convert_gp_costs.R

```diff
@@ -0,0 +1,143 @@
+#####################################################
+# Costs - GP Out of Hours
+# Author: Catherine Holland
+# Date: February 2022
+# Written on RStudio Server
+# Version of R - 3.6.1
+# Input -
+# Description -
+#####################################################
+
+
+## update file ##
+
+## Attendances taken from 2018 Primary Care Out of Hours Report ##
+# https://publichealthscotland.scot/publications/out-of-hours-primary-care-services-in-scotland/
+
+## Costs taken from R520 (Costbook) report for 2015/16 ##
+# https://beta.isdscotland.org/topics/finance/costs/ (R520)
+
+## The above should be checked / added to the Excel file 'OOH_Costs.xlsx' before running this syntax.
+
+
+library(dplyr)
+library(ggplot2)
+
+
+## data ##
+gp_out_of_hours <- readxl::read_xlsx(paste0(get_slf_dir(), "/Costs/OOH_Costs.xlsx"))
+
+
+## data - wide to long ##
+gp_out_of_hours <-
+  gp_out_of_hours %>%
+  pivot_longer(
+    contains("Consultations"),
+    names_to = "Year_Consultations", values_to = "Consultations"
+  ) %>%
+  pivot_longer(
+    contains("Cost"),
+    names_to = "Year_Cost", values_to = "Cost"
+  )
+
+
+## create year variable ##
+gp_out_of_hours <-
+  gp_out_of_hours %>%
+  mutate(Year = substr(Year_Consultations, 1, 4))
+
+
+## create cost per consultation ##
+gp_out_of_hours <-
+  gp_out_of_hours %>%
+  mutate(
+    cost_per_consultation = Cost * 1000 / Consultations
+  )
+
+
+## add in years by copying the most recent year ##
+year <- 1920
+
+tempyear1 <- paste0(as.numeric(substr(year, 1, 2)) + 1, as.numeric(substr(year, 3, 4)) + 1)
+tempyear2 <- paste0(as.numeric(substr(year, 1, 2)) + 2, as.numeric(substr(year, 3, 4)) + 2)
+
+
+
+## increase by 1% for every year after the latest ##
+gp_out_of_hours <-
+  gp_out_of_hours %>%
+  mutate(
+    cost_per_consultation =
+      if_else(Year > year,
+        cost_per_consultation * 1.01,
+        cost_per_consultation
+      )
+  ) %>%
+  # arrange by HB2019 and year
+  arrange(HB2019, Year)
```

Comment:
The idea here is to copy costs forward to years where we don't have them. Your code has two issues: 1) it doesn't create any new rows. 2) It would only increment costs by 1%, not 1% per year.

Reply:
We could just add 3 years extra as standard, something like.

```r
gp_out_of_hours %>%
  bind_rows(
    filter(., Year == latest_year) %>%
      mutate(
        cost_per_consultation = cost_per_consultation * (1.01)^1,
        Year = convert_year_to_fyyear(as.numeric(convert_fyyear_to_year(Year)) + 1)
      ),
    filter(., Year == latest_year) %>%
      mutate(
        cost_per_consultation = cost_per_consultation * (1.01)^2,
        Year = convert_year_to_fyyear(as.numeric(convert_fyyear_to_year(Year)) + 2)
      ),
    filter(., Year == latest_year) %>%
      mutate(
        cost_per_consultation = cost_per_consultation * (1.01)^3,
        Year = convert_year_to_fyyear(as.numeric(convert_fyyear_to_year(Year)) + 3)
      )
  )
```
Maybe you can think of a better way of doing this! It doesn't matter if we have 'too many' costs i.e. having 22/23 costs before 22/23 isn't an issue, but having costs missing is.

Reply:
potential fix using `map`

Reply:
Yeah, I thought of `map` or `reduce` but couldn't figure out how to use them for this...

I've done a vaguely similar thing in a different project with population files (copying population estimates to 'future' years), I just checked and I used a while loop for that
```r
copy_pops_for_additional_years <- function(pops_df, max_pop_year) {

  # Get the current max pop available
  max_pop_available <- pops_df %>%
    pull(year) %>%
    max()

  # Count how many years we are adding
  years_added <- 0

  # This will add on as many years as needed
  while (max(pops_df$year) < max_pop_year) {
    copied_pops <- pops_df %>%
      dplyr::filter(year == max(pops_df$year)) %>%
      dplyr::mutate(year = year + 1)

    pops_df <- dplyr::bind_rows(pops_df, copied_pops)

    years_added <- years_added + 1
  }

  message(paste(
    "*** Latest population year available:",
    max_pop_available,
    "This was copied",
    years_added,
    "time(s), we now have populations up to",
    max_pop_year,
    "***"
  ))

  return(pops_df)
}
```